src/scripting/lua_widget_methods.cpp

```diff
@@ -210,6 +210,45 @@ static int intf_find_widget(lua_State* L)
 
 namespace
 {
+	int number_of_items(gui2::listbox& mp)
+	{
+		return mp.get_item_count();
+	}
+	int number_of_items(gui2::multi_page& mp)
+	{
+		return mp.get_page_count();
+	}
+
+	int number_of_items(gui2::tree_view_node& mp)
+	{
+		return mp.count_children();
+	}
+
+	int number_of_items(gui2::tree_view& mp)
+	{
+		return number_of_items(mp.get_root_node());
+	}
+
+	// Checks, and makes sure that a (lua. 1-based) index to a container widget is valid
+	// The caller still have to subtract 1 when passing the index to c++
+	// (0-based) functions
+	template<typename TWidget>
+	void fix_index(lua_State* L, int arg, TWidget& w, bool for_insertion, int& index)
+	{
+		int nitems = number_of_items(w);
+		// index == nitems + 1 -> insert at the end.
+		int max = for_insertion ? nitems + 1 : nitems;
+
+		// 0 means insert at end.
+		if(index == 0) {
+			index = nitems + 1;
+		}
+		// todo: does a "negative indicies to "
```

Comment:
If index is negative, just add max+1 to it. It could _still_ be out of range after doing that, however.

I think an out-of-range index on insert should also be clipped to max instead of being an error.